### ISSUE TITLE
Add a better push implementation

### ISSUE THREAD

**Issue by elmattic** [elmattic/navi#64]
The code could be much much simpler.
And it won't work well on AVX2 path.

My idea is to broadcast the size of the mask plus as special value taken so that when you add with saturation you hit `0xff` for the bytes you want to keep.
Then you just need a lane id vector, add saturate with it. Last step you clear the non-saturated values with an all ones vector.
This vector full of ones is easily obtained by comparing any register with itself (zero and one idioms are virtually 'free').

**Comment by elmattic** [comment]
For some unknown to me reason, the intrinsics `_mm_set_epi32` and `_mm_set_epi16` are not exposed in `clang`.
That means a roundtrip to `eax` register which could be avoided.

Two solutions:

- inline assembly insanity: `__asm__("movd %1, %0" : "=x" (lo) : "m" (src));`

- copy paste clang trunk code, since it is exposed there (I've just checked on godbolt)

**Comment by elmattic** [comment]
Actually the code is currently faster, so I'll keep it simple.

**Comment by elmattic** [comment]
So my code works perfectly on my laptop but fail to pass the tests on CircleCI...

**Comment by elmattic** [comment]
So it breaks with O3 and O1, probably O2 too but success in O0, that means the compiler did something wrong, or maybe be I don't want the compiler always.
What is sure is that it's related to the PUSH16-32.
My first guess is to say that the undefined intrinsic wrecked it. So I will try to remove that.

**Comment by elmattic** [comment]
Hey... There are AVX-512 extensions on CircleCI processors!

Good to know because it's still not mainstream.
That being said I don't think it would be helpful in the context of the EVM (and AVX2 still waiting to be implemented and show benefits).

**Comment by elmattic** [comment]
So `clang 3.9.1` (CircleCI version) generates following optimized code for this two C functions:

```
        xorps   xmm0, xmm0
        ret
```

```
static __m128i all_ones()
{
  const __m128i undef = _mm_undefined_si128();
  return _mm_cmpeq_epi32(undef, undef);
}
```

```
static __m128i all_ones()
{
  const __m128i undef;
  return _mm_cmpeq_epi32(undef, undef);
}
```

It has been fixed starting from `clang 5.0.1`. And so my Apple's flavor of `clang` is probably based on, or contains a special fix for this bug, and then thank you.

**Comment by elmattic** [comment]
1296bb84221cc4885116db189e38decf377f57c4
